### PR TITLE
Tracelogging

### PR DESCRIPTION
**Pull request by jjhegedus**
I think this is pretty solid now.  The only thing I haven't done is set up tests.  The only way I can see to test it would be to write a consumer of trace log messages, start the consumer prior running the test and use the consumer to pass data back to the test.  It seems like more work than its worth.  I'm looking forward to seeing what you think.

### CODE REVIEW

File: tracelogging/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+if(${CMAKE_VERSION} VERSION_LESS 3.12)
+    cmake_policy(VERSION ${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION})
+endif()
+
+# detect if g3log_tracelogging is being included as a subdirectory
+if(NOT DEFINED PROJECT_NAME)
+  set(NOT_SUBPROJECT ON)
+endif()
+
+#list(APPEND CMAKE_CXX_SOURCE_FILE_EXTENSIONS c)
```
Comment (KjellKod) [inline tracelogging/CMakeLists.txt]: please remove dead code

File: README.md [hunk 2]
```diff
@@ -50,6 +50,36 @@ been tested on Cygwin.
 This component license is public domain, a.k.a the  UNLICENSE.
 See details at the sink [location](https://github.com/KjellKod/g3sinks/tree/master/logrotate).
 
+## g3sinks/tracelogging
+sink to allow logging through Windows TraceLogging
+
+### Building and installing g3log_tracelogging
```
Comment (KjellKod) [inline README.md]: how about this?
Keep this (or something like it)
```
## g3sinks/tracelogging
Sink to allow logging through Windows TraceLogging. 
For build instructions please see [tracelogging/README](tracelogging/README.md)
```
Remove the rest in this readme file. 

That way you don't have duplicated any info and won't have an issue with it being out of sync between the readmes.

File: tracelogging/include/g3log_tracelogging.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+TRACELOGGING_DECLARE_PROVIDER(traceProvider);
+
+#define TraceLogWrite(eventName, ...) \
+    _TlgWrite_imp(_TlgWrite, \
+    traceProvider, eventName, \
+    (NULL, NULL), \
+    __VA_ARGS__)
+
+// *** BEGIN EXAMPLE CLIENT TRACELOGGING DEFINITION ***
```
Comment (KjellKod) [inline tracelogging/include/g3log_tracelogging.h]: For readability I would suggest keeping the example within 
```
/** 

*/
```

And removing the `//`

File: tracelogging/example/src/tracelogging_example.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#include "Windows.h"
+#include <g3log/g3log.hpp>
+#include <g3log/logworker.hpp>
+#include <g3log_tracelogging/g3log_tracelogging.h>
+#include <TraceLoggingProvider.h>
```
Comment (KjellKod) [inline tracelogging/example/src/tracelogging_example.cpp]: Is `TraceLoggingProvider.h` available on Windows by default?
If so, would you be interested in working with me to get this code working under a cloud CI/Windows?
I.e similar to how g3log PRs work now: **https://github.com/KjellKod/g3log/blob/master/appveyor.yml**
(it should be done in a different PR for sure ... but would be absolutely kick ass if we could get this under CI build coverage )

File: tracelogging/CMakeLists.txt (hunk 1 above)
Comment (jjhegedus) [reply tracelogging/CMakeLists.txt]: Good catch.  I removed the dead code

File: README.md (hunk 2 above)
Comment (jjhegedus) [reply README.md]: Good idea.  I made the change you suggested

File: tracelogging/include/g3log_tracelogging.h (hunk 3 above)
Comment (jjhegedus) [reply tracelogging/include/g3log_tracelogging.h]: Another good suggestions.  I made the change.

File: tracelogging/example/src/tracelogging_example.cpp (hunk 4 above)
Comment (jjhegedus) [reply tracelogging/example/src/tracelogging_example.cpp]: Yes, I'm pretty sure it is available on Windows by default.  Yes, I would definitely be interested in working with you to get it working in CI.  I'm completely new to appveyor but I'll take a look and see what I can figure out.

**Comment by jjhegedus** [comment]
I've looked through some of the Appveyor docs and some examples, including the g3log example.  I think I understand how to get the build working through Appveyor.  I also think I understand how to run the example.  Unfortunately, since I don't have a test, all we could do is verify that everything builds and runs, not that it actually works.  I'm not sure how you feel about that.

Should we finish this PR out and then I open a new one with this completed PR as the baseline?

**Comment by jjhegedus** [comment]
By the way, thanks for the kind words.  I'm very excited that you liked what I did.  This is my first PR ever and I'll be very proud if it is accepted.  It feels like I'm finally joining the open source community and I could never have done it without your guidance.

**Comment by KjellKod** [comment]
Ill re-review the PR tomorrow. 

Yes. Let’s keep this PR focused and get another one up for the CI later.

**Comment by KjellKod** [comment]
hi @jjhegedus 
What's a good email to reach you at?  I've some questions about the tracelogging sink.

**Pull request merged**